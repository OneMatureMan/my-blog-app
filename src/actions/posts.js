import { v1 as uuidv1 } from 'uuid';
import moment from 'moment';

export const addPost = (
    {
        title = '',
        body = '',
        createdAt = 0
    } = {}
) => ({
    type: 'ADD_POST',
    post : {
        id: uuidv1(),
        title,
        body,
        createdAt: moment()
    }
})

export const editPost = (id, updates) => ({
    type: 'EDIT_POST',
    id,
    updates
});


export const removePost = ({ id } = {}) => ({
    type: 'REMOVE_POST',
    id
})