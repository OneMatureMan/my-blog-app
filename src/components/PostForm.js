import React from 'react';


class PostForm extends React.Component {
    state = {
        title: this.props.post ? this.props.post.title : '',
        body: this.props.post ? this.props.post.body : '',
        error: ''
    }


    handleTitleChange = (e) => {
        const title = e.target.value;
        this.setState(({ title }))
    }
    
    handleBodyChange = (e) => {
        const body = e.target.value;
        this.setState(({ body }))
    }

    onSubmit = (e) => {
        e.preventDefault();
        if (!this.state.title || !this.state.body) {
            return this.setState(({ error: 'Please enter both title and body'}))
        }
        this.setState(({ error: '' }))
        this.props.onSubmit({
            title: this.state.title,
            body: this.state.body
        })        
    }
    
    render() {
        return (
            <form onSubmit={this.onSubmit}>
                {this.state.error && <p>{this.state.error}</p>}
                <input 
                    type="text"
                    name='title'
                    placeholder='Title'
                    autoFocus
                    value = {this.state.title}
                    onChange={this.handleTitleChange}
                />
                <textarea
                    name='body'
                    placeholder='Enter the blogs body here'
                    value= {this.state.body}
                    onChange={this.handleBodyChange}
                />
                <button>Save Post</button>
            </form>
        )
    }
}


export default PostForm;