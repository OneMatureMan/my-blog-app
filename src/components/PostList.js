import React from 'react';
import { connect } from 'react-redux'
import PostListItem from './PostListItem';
import selectPosts from '../selectors/posts';

export const PostList = ({posts}) => (
    <div>
        <h1>THE POSTS!</h1>
        {posts.map((post) => {
            return <PostListItem key={post.id} {...post} />
        })}
    </div>
)

const mapStateToProps = state => ({
    posts: selectPosts(state.posts,state.filters)
})

export default connect(mapStateToProps)(PostList);


