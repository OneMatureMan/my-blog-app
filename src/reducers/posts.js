const postsDefaultState = []

const postsReducer = (state = postsDefaultState, action) => {
    switch (action.type) {
        case 'ADD_POST':
            return state.concat({ id:action.id, ...action.post});
        case 'EDIT_POST':
            return state.map(post => {
                if (post.id === action.id) {
                    return {
                        ...post,
                        ...action.updates
                    }
                } else {
                    return post;
                }
            })
        case 'REMOVE_POST':
            return state.filter(({ id }) => id !== action.id)
        default :
            return state;
    }
}

export default postsReducer;