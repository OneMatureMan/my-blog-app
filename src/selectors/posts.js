// get visible posts
export default (posts, { text, sortBy }) => {
    return posts.filter(({ title }) => title.toLowerCase().includes(text.toLowerCase()))
            .sort((a,b) => {
                if (sortBy === 'date') {
                    if (a.createdAt < b.createdAt) {
                        return 1;
                    }
                    if (a.createdAt > b.createdAt) {
                        return -1;
                    }
                    return 0;
                } else if (sortBy === 'title') {
                    
                    if (a.title.toLowerCase() < b.title.toLowerCase()) {
                        return -1;
                    }
                    
                    if (a.title.toLowerCase() > b.title.toLowerCase()) {
                        return 1;
                    }
                    return 0;
                }            
            })
}