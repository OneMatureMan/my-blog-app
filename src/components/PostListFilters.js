import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import { filterByText, sortByDate, sortByTitle } from '../actions/filters';


export class PostListFilters extends React.Component {
    onTextChange = (e) => {
        this.props.filterByText(e.target.value);
    }

    onSortByChange = (e) => {
        if (e.target.value === 'date'){
            return this.props.sortByDate()
        }
        return this.props.sortByTitle()
    }
    render() {
        return (
            <div>
                <input 
                    className='text-input'
                    placeholder='Search posts'
                    type="text"
                    value={this.props.filters.text}
                    onChange={this.onTextChange} 
                />
                <select 
                    className='select'
                    value={this.props.filters.sortBy}
                    onChange={this.onSortByChange}
                >
                    <option value="date">Date</option>
                    <option value="title">Title</option>
                </select>
                <Link to='/create'>
                    <button>Add Post</button>
                </Link>
            </div>
        )
    }
}

const mapStateToProps = state => ({
    filters: state.filters
})

const mapDispatchToProps = (dispatch) => ({
    filterByText: (text) => dispatch(filterByText(text)),
    sortByDate: () => dispatch(sortByDate()),
    sortByTitle: () => dispatch(sortByTitle())
});



export default connect(mapStateToProps, mapDispatchToProps)(PostListFilters);