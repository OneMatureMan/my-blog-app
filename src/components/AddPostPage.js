import React from 'react';
import { connect } from 'react-redux';
import PostForm from './PostForm';
import { addPost } from '../actions/posts';


export class AddPostPage extends React.Component {

    onSubmit = (expense) => {
        this.props.addPost(expense)
        this.props.history.push('/dashboard');
    }

    render() {
        return (
            <PostForm 
                onSubmit={this.onSubmit}
            />
        );
    };
};


const mapDispatchToProps = (dispatch) => ({
    addPost : (post) => dispatch(addPost(post))
})

export default connect(undefined, mapDispatchToProps)(AddPostPage);