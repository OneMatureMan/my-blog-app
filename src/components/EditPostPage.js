import React from 'react';
import { connect } from 'react-redux';
import { editPost, removePost } from '../actions/posts';
import PostForm from './PostForm';


export class EditPostPage extends React.Component {

    onSubmit = (post) => {
        this.props.editPost(this.props.post.id,post);
        this.props.history.push('/dashboard');
    }

    onRemovePost = () => {
        this.props.removePost({ id: this.props.post.id});
        this.props.history.push('/dashboard');
    }

    render() {
        return (
            <div>
                <PostForm 
                    onSubmit={this.onSubmit}
                    post={this.props.post}
                />
                <button onClick={this.onRemovePost}>Remove Post</button>
            </div>
        );
    };
};

const mapStateToProps = (state,props) => ({
    post: state.posts.find(post => post.id === props.match.params.id)   
})


const mapDispatchToProps = (dispatch) => ({
    editPost : (id, updates) => dispatch(editPost(id, updates)),
    removePost: (id) => dispatch(removePost(id))
})

export default connect(mapStateToProps, mapDispatchToProps)(EditPostPage);