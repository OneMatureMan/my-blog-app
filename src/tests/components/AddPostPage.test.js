import React from 'react';
import { shallow } from 'enzyme';
import { AddPostPage } from '../../components/AddPostPage';
import posts from '../fixtures/posts';

let addPost, history, wrapper;

beforeEach(() => {
    addPost = jest.fn();
    history = { push : jest.fn()};
    wrapper = shallow(<AddPostPage addPost={addPost} history={history} />)
})

test('should load Add post page component correctly', () => {
    expect(wrapper).toMatchSnapshot();
});

test('should handle onSubmit correctly', () => {
    wrapper.find('PostForm').prop('onSubmit')(posts[0]);
    expect(addPost).toHaveBeenLastCalledWith(posts[0]);
    expect(history.push).toHaveBeenLastCalledWith('/dashboard');
})