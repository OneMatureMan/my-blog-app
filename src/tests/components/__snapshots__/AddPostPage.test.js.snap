// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should load Add post page component correctly 1`] = `
<PostForm
  onSubmit={[Function]}
/>
`;
